### ISSUE TITLE
Native Sign In

### ISSUE THREAD

**Issue by bdlukaa** [supabase/supabase-flutter#5]
# Feature request

## Is your feature request related to a problem? Please describe.

I want to sign in using a sign in provider, but the current implementation is not good. Launching the browser is not a good user experience.

## Describe the solution you'd like

Use the [`google_sign_in`](https://pub.dev/packages/google_sign_in) package to sign the user in using Google and [`sign_in_with_apple`](https://pub.dev/packages/sign_in_with_apple) to sign the user in using Apple.

Both of the packages return the user credentials, so, if I could sign in using the user credentials (just like Firebase does), it'd be great!

## Aditional context

If launching the brower will still be the current solution, we could have [in-app-handling](https://pub.dev/packages/url_launcher#browser-vs-in-app-handling).
By default, Android opens up a browser when handling URLs. You can pass `forceWebView: true` parameter to tell the plugin to open a WebView instead.

**Comment by himeshp** [comment]
+1, I have been waiting for this implementation.

**Comment by dshukertjr** [comment]
There needs to be some work done on the server side, but once that is done, we should be able to add this feature!
https://github.com/supabase/gotrue/issues/140

**Comment by bdlukaa** [comment]
Depend on https://github.com/supabase/gotrue-js/issues/169

**Comment by DanMossa** [comment]
Made a PR to bring this to Dart
https://github.com/supabase-community/gotrue-dart/pull/61

**Comment by nextdev1111** [comment]
When will it be implemented

**Comment by Aurion72** [comment]
> Hello, I've implemented Sign In with Apple in my App, but on iOS instead of opening the sign in sheet in the app, it opens safari with Sign In with Apple the then the user is redirect in the app, I'm calling the function like this:
> 
> supabase.auth.signInWithProvider( Provider.apple, options: AuthOptions( redirectTo: 'io.supabase.flutterquickstart://login-callback/'), );
> 
> Is it possible to open the sheet directly in the app and ket the user sign in, I've tried getting the credential and then passing to su-abase, but I always got an error of missing 'expires_in' parameters. Anyone has a solution?
> 
> I'm using su-abase 0.3.6 version on flutter

Hello, if you have any solutions for this i take it :)

**Comment by hmarat** [comment]
@dshukertjr hey :) Do you have even a little progress? We all waiting for it )

**Comment by dshukertjr** [comment]
Thanks everyone for waiting patiently for this feature. Security comes first at Supabase, and we are making sure this feature lands securely when it does. Because of that, it is taking time. 

Again, thank you all so much for the patients.

**Comment by hmarat** [comment]
Hey @dshukertjr I have too ask you again about approximate TIME of this feature, because we should decide start it with firebase or wait for supabase. Thank you!

**Comment by bdlukaa** [comment]
As far as I can tell (and by the context of this issue and the library itself), native sign in has already been achieved with `signInWithOpenIDConnect`. Yet, it needs some improvements and we'd like to provide a effortless solution for the developer, that's why this isn't marked as complete yet.

I also recall some issues with it on iOS. Currently, the bug-fixes are a priority, but we're looking forward to implement this feature for the stable 1.0.

**Comment by dshukertjr** [comment]
@bdlukaa Thanks for chiming in. 

@hmarat Sorry but unfortunately I'm afraid there is no definite timeline that we can share. You can try out [this](https://github.com/supabase-community/gotrue-dart/pull/61/files) method in the mean while.

**Comment by hmarat** [comment]
@dshukertjr I don't see it in [supabase_flutter](https://pub.dev/packages/supabase_flutter) :(

**Comment by dshukertjr** [comment]
@himeshp 
It is there, but not documented . You can see [this PR](https://github.com/supabase-community/gotrue-dart/pull/61/files) to see how it works, but there are still some missing pieces, so it is not yet fully supported. 

```dart
    Supabase.instance.client.auth.signIn(oidc: OpenIDConnectCredentials());
```

**Comment by hmarat** [comment]
@himeshp Oh, that's very good!
1. Does it mean that I should get for example idToken of user with google sign in, and just call that api with providers.google?
2. For both signup and sign in cases .signIn() method should be used?? 
Thanks in advance!

**Comment by hmarat** [comment]
I'm doing in that way but always get null 🤔

**Comment by hmarat** [comment]
![image](https://user-images.githubusercontent.com/55658812/185779840-405c92bd-a7b5-49a9-a587-f5af0b7b153f.png)
@dshukertjr  here is the log :)

**Comment by hmarat** [comment]
@dshukertjr do you have even rest api for creating user? So we can write some cloud function or do it in my backend?

**Comment by hmarat** [comment]
@bdlukaa maybe you can help me with openID api? Whatever I do, it returns null and does not create new users! The project's progress is stopped because of auth(

**Comment by dshukertjr** [comment]
@himeshp 

>do you have even rest api for creating user? So we can write some cloud function or do it in my backend?

[Here](https://github.com/supabase/gotrue-js/blob/rc/src/GoTrueAdminApi.ts#L126) is an api to create a user

**Comment by dshukertjr** [comment]
@ZetiMente The current implementation is half baked, and there needs to be some additional work to be done on the backend.

**Comment by devon2018** [comment]
This is something that really needs to be implemented soon. I have just had an app rejected because of this, see screenshot.

<img width="1292" alt="Screenshot 2022-10-19 at 21 21 18" src="https://user-images.githubusercontent.com/46196252/196795845-3ea6ba2d-443c-4af9-a6fb-416042fd4cc3.png">

**Comment by DanMossa** [comment]
@devon2018 A PR just got merged into togrue. Give it a little and it should work !

**Comment by wiverson** [comment]
@DanMossa I'm blocking on needing native sign in as well. Any more details on how to a) find b) try out the PR...?

**Comment by DanMossa** [comment]
This is what I have at this moment.
1. Sign in with Google working on Android and Web. ✅ 
2. Sign in with Google working on iOS. ⛔  Blocked due to https://github.com/google/GoogleSignIn-iOS/issues/135 and https://github.com/supabase-community/gotrue-dart/pull/68#pullrequestreview-965583082 .
3. Sign in with Apple working on Android, Web, or IOS. ⛔ Blocked until https://github.com/supabase/gotrue/pull/689 gets deployed to hosted Supabase users.

**Comment by dshukertjr** [comment]
@devon2018 Out of curiosity, which id provider do you have on your app? Google and Apple?

**Comment by devon2018** [comment]
@dshukertjr So I've got Facebook, Google and Apple Sign In all in my app.

**Comment by mohamad-jawad** [comment]
@devon2018 @DanMossa @wiverson @dshukertjr 
I figured out a way for native google sign in with iOS:

As mentioned by @DanMossa, google sign in is working properly with android and Web, However with iOS it is not, since a nonce is embedded in id_token, and we don't have access to this nonce in order to pass it to supabase, so supabase returns error.

To overcome this issue I have to do my own authentication to google using  [flutter_appauth](https://pub.dev/packages/flutter_appauth) package, and I specify the nonce as empty

### First Step - Authorization
```
final FlutterAppAuth _appAuth = const FlutterAppAuth();
final AuthorizationRequest authorizationRequest;
authorizationRequest = AuthorizationRequest(
     Config().GOOGLE_CLIENT_ID_IOS,
     Config().redirectUrl(),
     issuer: Config().GOOGLE_ISSUER, // https://accounts.google.com
     scopes: ['openid', 'profile', 'email'],
     nonce: "", // empty nonce
     additionalParameters: {'audience': Config().GOOGLE_CLIENT_ID_WEB}, //define the aud as your web client_id, so the generated id_token aud will be the web client_id in order for supabase to use it 
);

// Requesting the auth token and waiting for the response
final AuthorizationResponse? authResponse = await _appAuth.authorize(
     authorizationRequest,
);
```

### Second Step - Token
```
final TokenRequest tokenRequest;
tokenRequest = TokenRequest(
     Config().GOOGLE_CLIENT_ID_IOS,
     Config().redirectUrl(),
     issuer: Config().GOOGLE_ISSUER,
     scopes: ['openid', 'profile', 'email'],
     authorizationCode: authResponse?.authorizationCode,
     codeVerifier: authResponse?.codeVerifier,
     nonce: "",
     additionalParameters: {'audience': Config().GOOGLE_CLIENT_ID_WEB},
);
final TokenResponse? tokenResponse = await _appAuth.token(
     tokenRequest,
);
String idToken = tokenResponse?.idToken ?? "";
```

### Final Step - Get your supabase account

```
backend.OpenIDConnectCredentials oidc = backend.OpenIDConnectCredentials(
     idToken: idToken,
     clientId: Config().GOOGLE_CLIENT_ID_IOS,
     provider: backend.Provider.google,
     issuer: Config().GOOGLE_ISSUER,
     nonce: "",
);
final response = await _supabaseClient.auth.signIn(oidc: oidc);
```

I hope it helps ♥️

**Comment by DanMossa** [comment]
@mohamad-jawad Great job! It's definitely a solid way of solving it in the mean time!

**Comment by devon2018** [comment]
@DanMossa @mohamad-jawad any idea what the equivalent of this `supabaseClient.auth.signIn(oidc: oidc);` would be in v1 of the library, signIn method no longer exists and i cant see anything in the gotrue lib that allows oidc.

**Comment by dshukertjr** [comment]
@devon2018 We have removed the open id connect signin option from both supabase-js v2 and supabase-flutter v1 until we have a solid implementation of it. Our auth team has a lot of feature requests on their plates, so if we could wait patiently until we come out with an official announcement of the feature, it would be great!

**Comment by devon2018** [comment]
@dshukertjr I release i'm being a pain lol, but is there a rough timeline for this one. I have a client app that needs releasing soon and I really love supabase thats why i suggested to the client we use it but this is a mega blocker, so will need to find another solution if I can't get social sign in working in the next few weeks.

**Comment by dshukertjr** [comment]
@devon2018 
I hear you. Thanks for the love for Supabase. 

Good news is that I found an official comment from the auth team just posted a few hours ago moving this in a positive direction! I will work closely with the supabase-js team to bring back the open id connect feature to the flutter library.

https://github.com/supabase/gotrue/issues/434#issuecomment-1287755049

> Hey everyone, the team and I discussed this at length. Here's our conclusions:
> 
> We recognize this is being used in the wild and is a hard requirement on iOS.
> The current implementation https://github.com/supabase/gotrue/issues/412 and we don't want to encourage people to be using something that is of low quality.
> Mid-term we do want to add official support for an OIDC login flow (that uses ID tokens), but this is unlikely to be backward compatible with what we have today.
> Thus we decided to do the following:
> 
> Add back support for this flow in v2 of gotrue-js, but under an @experimental annotation. We're going to be allocating some time to do this in the next 2 to 3 weeks.
> Since Sign-in with Apple is a hard requirement in mobile apps on iOS, we'll try to address some of the known issues in that flow only. Since we're busy with some other priorities, we can't promise a timeline on addressing this.
> Any PRs from the community on this topic will be considered closely, in an effort to unblock you.
> Once we address OIDC ID token login flows properly, the existing flow will be deprecated and the existing API is likely to be removed with some prior notice. Prepare your codebase for such an eventuality in the future.
> Feel free to let us know how you feel about this or if we should reconsider something. Thank you for your dedication so far!

**Comment by DanMossa** [comment]
@dshukertjr Okay cool. So at this point, what exactly is happening lol.

We removed OIDC from flutter because it was removed from JS. The Supabase team not wants to keep this version of OIDC and is keeping it as a legacy feature until they come up with a good implementation. So this Flutter package will bring back the code that was removed and mark it as deprecated, yeah?

**Comment by dshukertjr** [comment]
We will basically mirror what they do on gotrue-js.

**Comment by thilojaeggi** [comment]
Any update on this, I don't see any way to implement native iOS sign in right now?

**Comment by oliverbytes** [comment]
My app just got rejected because supabase-flutter opens the OAuth screen via an external browser. Full rejection message below:

![NexSnap-Dec-15-2022_06-07-05_AM](https://user-images.githubusercontent.com/396366/207725431-313a2ac9-8515-4b31-8f7f-382694956c95.png)

**Comment by ZetiMente** [comment]
Oh great. Now we have to worry about App Store rejection...

**Comment by DanMossa** [comment]
I'm using native sign in with Apple on iOS and native sign in with Google on Android.

Let me know if you guys want a tutorial and if I find time for it I can do so. But my current limitation is that I can't have sign in with Google on iOS and vice versa.

**Comment by oliverbytes** [comment]
> I'm using native sign in with Apple on iOS and native sign in with Google on Android.
> 
> Let me know if you guys want a tutorial and if I find time for it I can do so. But my current limitation is that I can't have sign in with Google on iOS and vice versa.

Please do share your workaround on this. It would help a lot. Thank you

**Comment by wiverson** [comment]
> I'm using native sign in with Apple on iOS and native sign in with Google on Android.
> 
> Let me know if you guys want a tutorial and if I find time for it I can do so. But my current limitation is that I can't have sign in with Google on iOS and vice versa.

Even just a basic high level approach would be helpful...

**Comment by mohsin2596** [comment]
> I'm using native sign in with Apple on iOS and native sign in with Google on Android.
> 
> Let me know if you guys want a tutorial and if I find time for it I can do so. But my current limitation is that I can't have sign in with Google on iOS and vice versa.

We are also stuck on this, would be great help if you could share something

**Comment by thilojaeggi** [comment]
@DanMossa
Facing the same issue, are you willing to share?

**Comment by DanMossa** [comment]
I was about to write up a quick guide on how I did it, but this PR broke Supabase's native sign in. :/
https://github.com/supabase/gotrue/pull/927

**Comment by DanMossa** [comment]
I'm working on a PR to add Native sign in to flutter supabase!

You can take a look here at an example WIP doc!

https://github.com/supabase/gotrue-dart/pull/119

**Comment by habibasseiss** [comment]
Awesome @DanMossa! Now that [supabase/gotrue-dart/pull/119](https://github.com/supabase/gotrue-dart/pull/119) is merged, what would be the next steps so we can use native auth in a Flutter app?

**Comment by theniceboy** [comment]
It turns out this **won't work** if you have have apps **on more than one platform** because in the current supabase dashboard you cannot configure google sign-in to use more than more OAuth Client ID.
![IMG_4541AC9BD04C-1](https://user-images.githubusercontent.com/8187501/218244210-0ba40ba3-23e7-4926-8aa4-b03056da6597.jpeg)

**Comment by alexrabin** [comment]
> It turns out this **won't work** if your app is **on more than one platform** because in the current supabase dashboard you cannot configure google sign-in to use more than more OAuth Client ID.
> ![IMG_4541AC9BD04C-1](https://user-images.githubusercontent.com/8187501/218244210-0ba40ba3-23e7-4926-8aa4-b03056da6597.jpeg)

Same problem with Sign in with Apple for android and web
![Screenshot 2023-02-14 at 8 27 49 PM](https://user-images.githubusercontent.com/15949910/218919902-62c1e773-4ecd-4c14-8cff-87a0b08eefd8.png)

**Comment by DanMossa** [comment]
I've been busy, but I'll get to working on that document sooner than later. This feature absolutely works and I'm actively using it. There are some limitations with the current implementation but it works.

**Comment by theniceboy** [comment]
@DanMossa Can you maybe provide a quick high level description on your solution, please? Thank you so much!

**Comment by theniceboy** [comment]
@DanMossa 

Actually nevermind... I saw your [comment](https://github.com/supabase/gotrue-dart/pull/119#issuecomment-1432207085) on #119 

Sign-in with Google will work if you're okay with having none of your other platforms use it. If you implement sign-in with google on iOS, your other apps (Android, web, desktop) can't use sign-in with google.

See reply https://github.com/supabase/gotrue-dart/pull/119#issuecomment-1432244047

> Currently If you configure any platform to use sign-in with google, you will be using the OAuth client ID for the web, because the current Supabase sign-in flow always uses a browser. If you want to use this native sign-in feature right now on let's say iOS, that means you have to set the OAuth client ID in your Supabase dashboard to use the iOS client ID. If you do that, native sign-in with Google will indeed work on your iOS app, however, none of your other platforms, such as Web, Android, or Desktop, can even implement sign-in with Google, because the current Supabase backend only allows one client ID per auth provider.

**Comment by alexrabin** [comment]
I'm in the exact same boat as @theniceboy. I need to be able to sign in with the same provider on multiple platforms.

**Comment by avalanche-tm** [comment]
Is this gonna be fixed or we'll be left with incomplete solution?

**Comment by DanMossa** [comment]
I'd like for us to track progress on this new ticket.
https://github.com/supabase/supabase-flutter/issues/399
I want to keep the top comment always updated with the status of different solutions being resolved.
@bdlukaa @dshukertjr

**Comment by dshukertjr** [comment]
Thanks @DanMossa for creating a issue with all the details. 

I just wanted to share here that the issue of not being able to sign in via Google on all web, android, and iOS is being addressed with the auth team, and we will have a solution for it.  

cc @alexrabin @theniceboy

**Comment by MichelBusse** [comment]
I had the same issue with my app getting rejected like @oliverbytes. Since I am developing with flutter, the easiest solution for now seemed adding an extension to the GoTrueClient and implementing an alternative method for signInWithOAuth with a few tweaks. I just launched the url in an InAppWebView  with url_launcher, but you could also add extensive Web-View support to fit your needs. Just check the code below:

```dart
extension GoTrueClientSignInProvider on GoTrueClient {
  Future<bool> signInWithOAuthInWebView(
      Provider provider, {
        String? redirectTo,
        String? scopes,
        Map<String, String>? queryParams,
      }) async {
    final res = await getOAuthSignInUrl(
      provider: provider,
      redirectTo: redirectTo,
      scopes: scopes,
      queryParams: queryParams,
    );
    final url = Uri.parse(res.url!);
    final result = await launchUrl(
      url,
      mode: LaunchMode.inAppWebView,
      webOnlyWindowName: '_self',
    );
    return result;
  }
}
```

**Comment by dshukertjr** [comment]
Wanted to share [the news](https://supabase.com/blog/supabase-auth-sso-pkce#native-apple-login-on-ios) here we now have a method to perform native sign in with Apple. 

```dart
final AuthResponse response = await supabase.auth.signInWithApple();
```
Huge shout out to @DanMossa for all the hard work he put into bringing this feature!

We are working on bringing native Google login on multiple platforms at the moment, and hope to share the good news there soon.

Once we add support for Google login on multiple platform, we will create a `signInWithGoogle()` method. With an implementation using [AppAuth](https://pub.dev/packages/flutter_appauth) package.

<details>
  <summary>Native Google login code</summary>
```dart
import 'dart:convert';
import 'dart:math';

import 'package:crypto/crypto.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter_appauth/flutter_appauth.dart';

Future<AuthResponse> signInWithGoogle(String clientId) {
  // Just a random string
  final rawNonce = _generateRandomString();
  final hashedNonce =
      sha256.convert(utf8.encode(rawNonce)).toString();
    
  /// bundle ID of the app
  const bundleId = 'com.supabase.authflow';
  
  /// fixed for google login
  // const redirectUrl = '$bundleId:/google_auth';
  const redirectUrl = '$bundleId:/google_auth';
  
  /// fixed for google login
  const discoveryUrl =
      'https://accounts.google.com/.well-known/openid-configuration';
  
  // authorize the user by opening the concent page
  final result = await appAuth.authorize(
    AuthorizationRequest(
      clientId,
      redirectUrl,
      discoveryUrl: discoveryUrl,
      nonce: hashedNonce,
      scopes: [
        'openid',
        'email',
      ],
    ),
  );
  
  if (result == null) {
    return;
  }
  
  // Request the access and id token to google
  final tokenResult = await appAuth.token(
    TokenRequest(
      clientId,
      redirectUrl,
      authorizationCode: result.authorizationCode,
      discoveryUrl: discoveryUrl,
      codeVerifier: result.codeVerifier,
      nonce: result.nonce,
      scopes: [
        'openid',
        'email',
      ],
    ),
  );
  
  final idToken = tokenResult?.idToken;
  
  if (idToken == null) {
    throw 'No idToken';
  }
  
  return supabase.auth.signInWithIdToken(
    provider: Provider.google,
    idToken: idToken,
    nonce: rawNonce,
  );
}
```
</details>

**Comment by alexrabin** [comment]
This is great! Does this also allow websites to Sign in with Apple as well?

**Comment by dshukertjr** [comment]
@alexrabin Sign in with Apple on a website was always supported using `signInWithOAuth()`, and still is the method to be used when signing in using Apple.

**Comment by oliverbytes** [comment]
The latest news is excellent for Supabase + Apple devs. Kudos @DanMossa 🤘. But unfortunately, it doesn't work for me (I tested iOS + macOS). I followed the setup and simple code to use. The native Apple sign-in dialog pops up but after successfully signing in, I get no response at all. No callbacks happen.

**Comment by dshukertjr** [comment]
@oliverbytes 
I was not able to reproduce the issue you are having on iOS simulator and macOS in debug mode. With the following code, I was able to confirm that both onAuthStateChange callback and the response from `signInWithApple` both works. At least you should see an error being thrown if something is not working.

```dart
supabase.auth.onAuthStateChange.listen((event) {
  print(event.event);
  print(event.session.toString());
});

final response = await supabase.auth.signInWithApple();
print(response.toString());
```

**Comment by oliverbytes** [comment]
> but after successfully signing in, I get no response at all. No callbacks happen to `onAuthStateChange` listener

Please ignore. I was dealing with a recently transferred app, that needed to update the client's secret keys and I guess it needed a bit of propagation time to make it finally work. Now it works on both macOS and iOS simulators 🤘 thanks a lot

**Comment by gregorym** [comment]
@dshukertjr Any update on when native signin will support Google?

**Comment by elliottetzkorn** [comment]
Google sign in is waiting on Google - https://github.com/google/GoogleSignIn-iOS/pull/244

Pretty frustrating that the code is ready to go, and they have been I guess ignoring it for months.. I guess tag this CR if you want to encourage them to prioritize it

**Comment by elliottetzkorn** [comment]
> @dshukertjr Any update on when native signin will support Google?

Google sign in is waiting on Google - https://github.com/google/GoogleSignIn-iOS/pull/244

Pretty frustrating that the code is ready to go, and they have been I guess ignoring it for months.. I guess tag this CR if you want to encourage them to prioritize it

**Comment by elliottetzkorn** [comment]
> @dshukertjr Any update on when native signin will support Google?

I think supabase is stuck until they prioritize this CR, so it could be many more months, or it could be done next week if they get to it..

**Comment by gregorym** [comment]
Is there some sort of work-around? Can we do the authentication manually from our server and sign-in the user?

**Comment by MexJason** [comment]
Hey I have been following this issue and im not too familiar with the backend, but I think this may be helpful for the Sign in with google on iOS side. 

https://developers.google.com/identity/sign-in/ios/backend-auth

**Comment by wiverson** [comment]
The flow is, unfortunately, kind of complicated.

- The heart of it is to get a JWT from Apple that is used as an authentication token. That JWT is generated via the native Sign in with Apple UI dialog.
- You will need an App ID and a Service ID from Apple. The App ID is what is used to flag the native iOS client. The Service ID is used by Supabase to validate a JWT from Apple via Apple's server-side APIs.
- You pass in something called a nonce (one time use string) to the native Sign in with Apple dialog. That nonce is embedded in to the JWT you get back from Apple. You then send the Apple JWT along with the nonce hash to Supabase. Supabase/GoTrue then validate that the Apple JWT and the nonce are valid, including checking the timestamp. The flow here is a bit funky IIRC, I think the nonce gets sent to Apple and comes back in the JWT straight and then you send a hash of the nonce to Supabase.

I made a video on all of this stuff working in a C#/Unity project, including helping out with getting the supabase-gotrue implementation to include the details cribbed from the Flutter implementation. Hopefully the video can help with sorting out the Sign in with Apple flow for Flutter as well. Note that I put links in the description for more info, including links to some gists.

https://youtu.be/S0hTwtsUWcM

**Comment by dshukertjr** [comment]
I am closing this issue as native sign-in with Apple and sign-in with Google have been added with multi-platform support. 

- [Native sign-in with Google setup video](https://youtu.be/YtvxRgGouwg)
- [Setup native sign-in with Apple with Supabase and Flutter](https://pub.dev/packages/supabase_flutter#native-apple-sign-in)
- [Setup native sign-in with Google with Supabase and Flutter](https://pub.dev/packages/supabase_flutter#native-google-sign-in)

**Comment by ozasadnyy** [comment]
Hey @dshukertjr, thanks for the video and article! Do you plan to support [google_sign_in](https://pub.dev/packages/google_sign_in) library in the future? It would save a lot of boilerplate.

**Comment by dshukertjr** [comment]
@ozasadnyy Hopefully yes, but the iOS version of Google sign in library has an issue where it does not allow the user to set their own nonce, which creates an security issue. Once this is fixed, developers can use the google sign in library with Supabase. 
https://github.com/google/GoogleSignIn-iOS/issues/135#issuecomment-1643759758

**Comment by dshukertjr** [comment]
@mohsin2596 
Hmm, from the error message, it seems like you haven't added your app's bundle ID in your Supabase dashboard under auth>providers>Apple, but could you double check it?

**Comment by mohsin2596** [comment]
@dshukertjr I can confirm that auth providers on dashboard are setup correctly. This error only popped up for me when I updated Supabase to the latest package and allowed users to use Sign in with Google along with Sign in with Apple

**Comment by dshukertjr** [comment]
@mohsin2596 From your Supabase dashboard, could you take a look at the auth logs under `logs>auth`? There you can see more detailed error messages typically.

**Comment by mohsin2596** [comment]
> @mohsin2596 From your Supabase dashboard, could you take a look at the auth logs under `logs>auth`? There you can see more detailed error messages typically.

```
Log ID
bf80622d-2619-425c-9949-7add576f669f

Log Timestamp (UTC)
2023-08-21T18:57:56.000Z

Log Event Message
{"component":"api","error":"invalid request: Unacceptable audience in id_token","level":"info","method":"POST","msg":"invalid request: Unacceptable audience in id_token","path":"/token","referer":"","remote_addr":"77.16.48.222","time":"2023-08-21T18:57:56Z","timestamp":"2023-08-21T18:57:55Z"}

Log Metadata
[
  {
    "message": null,
    "timestamp": "2023-08-21T18:57:55Z",
    "__MONOTONIC_TIMESTAMP": null,
    "CODE_FUNC": null,
    "instance_id": null,
    "status": null,
    "_CMDLINE": null,
    "method": "POST",
    "_SYSTEMD_CGROUP": null,
    "CODE_FILE": null,
    "EXECUTABLE": null,
    "_EXE": null,
    "UNIT": null,
    "level": "info",
    "_COMM": null,
    "duration": null,
    "issuer": null,
    "_LINE_BREAK": null,
    "_SOURCE_REALTIME_TIMESTAMP": null,
    "msg": "invalid request: Unacceptable audience in id_token",
    "action": null,
    "login_method": null,
    "_UID": null,
    "host": "db-erlrzretcsxoskhdifia",
    "PRIORITY": null,
    "_CAP_EFFECTIVE": null,
    "_PID": null,
    "INVOCATION_ID": null,
    "_SYSTEMD_UNIT": null,
    "source_type": null,
    "SYSLOG_FACILITY": null,
    "request_id": null,
    "CODE_LINE": null,
    "path": "/token",
    "component": "api",
    "project": null,
    "user_id": null,
    "auth_event": [],
    "args": [],
    "referer": null,
    "factor_id": null,
    "provider": null,
    "client_id": null,
    "remote_addr": "77.16.48.222",
    "_SYSTEMD_SLICE": null,
    "_SYSTEMD_INVOCATION_ID": null,
    "header": null,
    "_MACHINE_ID": null,
    "_AUDIT_LOGINUID": null,
    "_TRANSPORT": null,
    "_SELINUX_CONTEXT": null,
    "MESSAGE_ID": null,
    "__REALTIME_TIMESTAMP": null,
    "metadata": [],
    "_STREAM_ID": null,
    "metering": null,
    "time": null,
    "_GID": null,
    "_BOOT_ID": null,
    "SYSLOG_IDENTIFIER": null,
    "_AUDIT_SESSION": null,
    "error": "invalid request: Unacceptable audience in id_token"
  }
]
```

**Comment by mohsin2596** [comment]
Has anybody come across this issue? My app has been rejected by app review and I cant move forward without having it working

**Comment by dshukertjr** [comment]
@mohsin2596 
I have opened an issue on Gotrue repo. Our auth team will take a look at it soon.

**Comment by dshukertjr** [comment]
@mohsin2596 
Would you be able to submit a support ticket about this one? It might be a instance specific thing. 

https://supabase.com/dashboard/support/new

**Comment by mohsin2596** [comment]
@dshukertjr done. Will update here if it was instance related

**Comment by mohsin2596** [comment]
@dshukertjr After disabling and enabling the provider again it started working

**Comment by wiverson** [comment]
So, on Friday (apx 8/25) my Supabase instance started rejecting Sign In With Apple (native dialog). The relevant section in my logs was this one (redacted IP address):

```
{"component":"api","error":"invalid request: Unacceptable audience in id_token","level":"info","method":"POST","msg":"invalid request: Unacceptable audience in id_token","path":"/token","referer":"http://localhost:3000","remote_addr":"my ip address","time":"2023-08-29T18:21:59Z","timestamp":"2023-08-29T18:21:59Z"}
```

At first I thought this was an issue with Apple but it turned out to be Supabase. After a weekend of not having Sign in with Apple working I found this thread and flipping it off and back on again everything started working again.

Not sure what happened on the server side but ugh. :/

**Comment by dshukertjr** [comment]
@mohsin2596 @wiverson Thanks for reporting. We are looking into it.

**Comment by bitnom** [comment]
> @mohsin2596 @wiverson Thanks for reporting. We are looking into it.

Any update on this? Same error is happening when attempting to perform Google sign-in in Swift.

**Comment by mohsin2596** [comment]
> > @mohsin2596 @wiverson Thanks for reporting. We are looking into it.
> 
> Any update on this? Same error is happening when attempting to perform Google sign-in in Swift.

@bitnom Make sure you have comma separated values under Authorized Client IDs in Google Provider for both iOS and Android.

**Comment by wiverson** [comment]
FWIW no update on my end other than after flipping off & on again things still seem to be working.

**Comment by joaoantunes-mobiweb** [comment]
Hi,

I know this is a flutter thread but I am trying to implement the Native Google Sign in on react-native for IOS, and after getting the id token correctly ,I am having the same error mentioned here:

```json
{"data": {"session": null, "user": null}, "error": [AuthApiError: Unacceptable audience in id_token]}
```

While executing the following code:

```tsx
        const userInfo = await GoogleSignin.signIn();
        console.log('id token: ', userInfo.idToken);
        if (userInfo.idToken) {
          const {data, error} = await supabase.auth.signInWithIdToken({
            provider: 'google',
            token: userInfo.idToken,
          });

          console.log('Auth: ', {data, error});
        }
```

I am using [@react-native-google-signin/google-signin](https://github.com/react-native-google-signin/google-signin) library to get id token as suggested from the documentation [here](https://supabase.com/docs/guides/auth/social-login/auth-google#using-native-sign-in).

I have already checked that my bundleId (defined on XCode/Info.plist) matches with the one defined on Google Client Id, however, I keep having the same error. 

I am testing it in a iOS simulator locally, would it be a problem? I have no more clues how to debug it.

On my code everything is new since I have started working on this project during the last 3 weeks and I have only started using supabase at this point, since I am trying to make it the choice in my company for new projects, and this is an important feature. I can Google Sign In working with Web OAuth redirect, however, as it was already mentioned in this thread, on mobile it is important to have the native solution.

It would be great if I could have your help on debugging it. Thanks a lot!

I leave here the log on supabase -> auth:

```sh
Log ID
d1ff7039-9153-4adc-b3a4-340da0af989d

Log Timestamp (UTC)
2023-09-13T06:51:05.000Z

Log Event Message
{"component":"api","error":"invalid request: Unacceptable audience in id_token","level":"info","method":"POST","msg":"invalid request: Unacceptable audience in id_token","path":"/token","referer":"http://localhost:5173","remote_addr":"2.82.205.134","time":"2023-09-13T06:51:05Z","timestamp":"2023-09-13T06:51:05Z"}

Log Metadata
[
  {
    "message": null,
    "timestamp": "2023-09-13T06:51:05Z",
    "__MONOTONIC_TIMESTAMP": null,
    "CODE_FUNC": null,
    "instance_id": null,
    "status": null,
    "_CMDLINE": null,
    "method": "POST",
    "_SYSTEMD_CGROUP": null,
    "CODE_FILE": null,
    "EXECUTABLE": null,
    "_EXE": null,
    "UNIT": null,
    "level": "info",
    "_COMM": null,
    "duration": null,
    "issuer": null,
    "_LINE_BREAK": null,
    "_SOURCE_REALTIME_TIMESTAMP": null,
    "msg": "invalid request: Unacceptable audience in id_token",
    "action": null,
    "login_method": null,
    "_UID": null,
    "host": "db-nfdyvbqnfmamupnzagpf",
    "PRIORITY": null,
    "_CAP_EFFECTIVE": null,
    "_PID": null,
    "INVOCATION_ID": null,
    "_SYSTEMD_UNIT": null,
    "source_type": null,
    "SYSLOG_FACILITY": null,
    "request_id": null,
    "CODE_LINE": null,
    "path": "/token",
    "component": "api",
    "project": null,
    "user_id": null,
    "auth_event": [],
    "args": [],
    "referer": "http://localhost:5173",
    "factor_id": null,
    "provider": null,
    "client_id": null,
    "remote_addr": "2.82.205.134",
    "_SYSTEMD_SLICE": null,
    "_SYSTEMD_INVOCATION_ID": null,
    "header": null,
    "_MACHINE_ID": null,
    "_AUDIT_LOGINUID": null,
    "_TRANSPORT": null,
    "_SELINUX_CONTEXT": null,
    "MESSAGE_ID": null,
    "__REALTIME_TIMESTAMP": null,
    "metadata": [],
    "_STREAM_ID": null,
    "metering": null,
    "time": null,
    "_GID": null,
    "_BOOT_ID": null,
    "SYSLOG_IDENTIFIER": null,
    "_AUDIT_SESSION": null,
    "error": "invalid request: Unacceptable audience in id_token"
  }
]
```

Thanks!

**Comment by joaoantunes-mobiweb** [comment]
> Hi,
> 
> I know this is a flutter thread but I am trying to implement the Native Google Sign in on react-native for IOS, and after getting the id token correctly ,I am having the same error mentioned here:
> 
> ```json
> {"data": {"session": null, "user": null}, "error": [AuthApiError: Unacceptable audience in id_token]}
> ```
> 
> While executing the following code:
> 
> ```tsx
>         const userInfo = await GoogleSignin.signIn();
>         console.log('id token: ', userInfo.idToken);
>         if (userInfo.idToken) {
>           const {data, error} = await supabase.auth.signInWithIdToken({
>             provider: 'google',
>             token: userInfo.idToken,
>           });
> 
>           console.log('Auth: ', {data, error});
>         }
> ```
> 
> I am using [@react-native-google-signin/google-signin](https://github.com/react-native-google-signin/google-signin) library to get id token as suggested from the documentation [here](https://supabase.com/docs/guides/auth/social-login/auth-google#using-native-sign-in).
> 
> I have already checked that my bundleId (defined on XCode/Info.plist) matches with the one defined on Google Client Id, however, I keep having the same error.
> 
> I am testing it in a iOS simulator locally, would it be a problem? I have no more clues how to debug it.
> 
> On my code everything is new since I have started working on this project during the last 3 weeks and I have only started using supabase at this point, since I am trying to make it the choice in my company for new projects, and this is an important feature. I can Google Sign In working with Web OAuth redirect, however, as it was already mentioned in this thread, on mobile it is important to have the native solution.
> 
> It would be great if I could have your help on debugging it. Thanks a lot!
> 
> I leave here the log on supabase -> auth:
> 
> ```shell
> Log ID
> d1ff7039-9153-4adc-b3a4-340da0af989d
> 
> Log Timestamp (UTC)
> 2023-09-13T06:51:05.000Z
> 
> Log Event Message
> {"component":"api","error":"invalid request: Unacceptable audience in id_token","level":"info","method":"POST","msg":"invalid request: Unacceptable audience in id_token","path":"/token","referer":"http://localhost:5173","remote_addr":"2.82.205.134","time":"2023-09-13T06:51:05Z","timestamp":"2023-09-13T06:51:05Z"}
> 
> Log Metadata
> [
>   {
>     "message": null,
>     "timestamp": "2023-09-13T06:51:05Z",
>     "__MONOTONIC_TIMESTAMP": null,
>     "CODE_FUNC": null,
>     "instance_id": null,
>     "status": null,
>     "_CMDLINE": null,
>     "method": "POST",
>     "_SYSTEMD_CGROUP": null,
>     "CODE_FILE": null,
>     "EXECUTABLE": null,
>     "_EXE": null,
>     "UNIT": null,
>     "level": "info",
>     "_COMM": null,
>     "duration": null,
>     "issuer": null,
>     "_LINE_BREAK": null,
>     "_SOURCE_REALTIME_TIMESTAMP": null,
>     "msg": "invalid request: Unacceptable audience in id_token",
>     "action": null,
>     "login_method": null,
>     "_UID": null,
>     "host": "db-nfdyvbqnfmamupnzagpf",
>     "PRIORITY": null,
>     "_CAP_EFFECTIVE": null,
>     "_PID": null,
>     "INVOCATION_ID": null,
>     "_SYSTEMD_UNIT": null,
>     "source_type": null,
>     "SYSLOG_FACILITY": null,
>     "request_id": null,
>     "CODE_LINE": null,
>     "path": "/token",
>     "component": "api",
>     "project": null,
>     "user_id": null,
>     "auth_event": [],
>     "args": [],
>     "referer": "http://localhost:5173",
>     "factor_id": null,
>     "provider": null,
>     "client_id": null,
>     "remote_addr": "2.82.205.134",
>     "_SYSTEMD_SLICE": null,
>     "_SYSTEMD_INVOCATION_ID": null,
>     "header": null,
>     "_MACHINE_ID": null,
>     "_AUDIT_LOGINUID": null,
>     "_TRANSPORT": null,
>     "_SELINUX_CONTEXT": null,
>     "MESSAGE_ID": null,
>     "__REALTIME_TIMESTAMP": null,
>     "metadata": [],
>     "_STREAM_ID": null,
>     "metering": null,
>     "time": null,
>     "_GID": null,
>     "_BOOT_ID": null,
>     "SYSLOG_IDENTIFIER": null,
>     "_AUDIT_SESSION": null,
>     "error": "invalid request: Unacceptable audience in id_token"
>   }
> ]
> ```
> 
> Thanks!

After some more debugging and updating the **Client ID (for OAuth)** at Google Provider on Supabase Dashboard I started having the following error:

```json
{"data": {"session": null, "user": null}, "error": [AuthApiError: Passed nonce and nonce in id_token should either both exist or not.]}

Related log:

```sh
Log ID
0846b051-52b4-47a8-9c2c-e2e638fcf2be

Log Timestamp (UTC)
2023-09-13T13:25:24.000Z

Log Event Message
{"component":"api","error":"invalid request: Passed nonce and nonce in id_token should either both exist or not.","level":"info","method":"POST","msg":"invalid request: Passed nonce and nonce in id_token should either both exist or not.","path":"/token","referer":"http://localhost:5173","remote_addr":"2.82.205.134","time":"2023-09-13T13:25:24Z","timestamp":"2023-09-13T13:25:24Z"}

Log Metadata
[
  {
    "message": null,
    "timestamp": "2023-09-13T13:25:24Z",
    "__MONOTONIC_TIMESTAMP": null,
    "CODE_FUNC": null,
    "instance_id": null,
    "status": null,
    "_CMDLINE": null,
    "method": "POST",
    "_SYSTEMD_CGROUP": null,
    "CODE_FILE": null,
    "EXECUTABLE": null,
    "_EXE": null,
    "UNIT": null,
    "level": "info",
    "_COMM": null,
    "duration": null,
    "issuer": null,
    "_LINE_BREAK": null,
    "_SOURCE_REALTIME_TIMESTAMP": null,
    "msg": "invalid request: Passed nonce and nonce in id_token should either both exist or not.",
    "action": null,
    "login_method": null,
    "_UID": null,
    "host": "db-nfdyvbqnfmamupnzagpf",
    "PRIORITY": null,
    "_CAP_EFFECTIVE": null,
    "_PID": null,
    "INVOCATION_ID": null,
    "_SYSTEMD_UNIT": null,
    "source_type": null,
    "SYSLOG_FACILITY": null,
    "request_id": null,
    "CODE_LINE": null,
    "path": "/token",
    "component": "api",
    "project": null,
    "user_id": null,
    "auth_event": [],
    "args": [],
    "referer": "http://localhost:5173",
    "factor_id": null,
    "provider": null,
    "client_id": null,
    "remote_addr": "2.82.205.134",
    "_SYSTEMD_SLICE": null,
    "_SYSTEMD_INVOCATION_ID": null,
    "header": null,
    "_MACHINE_ID": null,
    "_AUDIT_LOGINUID": null,
    "_TRANSPORT": null,
    "_SELINUX_CONTEXT": null,
    "MESSAGE_ID": null,
    "__REALTIME_TIMESTAMP": null,
    "metadata": [],
    "_STREAM_ID": null,
    "metering": null,
    "time": null,
    "_GID": null,
    "_BOOT_ID": null,
    "SYSLOG_IDENTIFIER": null,
    "_AUDIT_SESSION": null,
    "error": "invalid request: Passed nonce and nonce in id_token should either both exist or not."
  }
]
```

Thanks!

**Comment by tobi404** [comment]
> So, on Friday (apx 8/25) my Supabase instance started rejecting Sign In With Apple (native dialog). The relevant section in my logs was this one (redacted IP address):
> 
> ```
> {"component":"api","error":"invalid request: Unacceptable audience in id_token","level":"info","method":"POST","msg":"invalid request: Unacceptable audience in id_token","path":"/token","referer":"http://localhost:3000","remote_addr":"my ip address","time":"2023-08-29T18:21:59Z","timestamp":"2023-08-29T18:21:59Z"}
> ```
> 
> At first I thought this was an issue with Apple but it turned out to be Supabase. After a weekend of not having Sign in with Apple working I found this thread and flipping it off and back on again everything started working again.
> 
> Not sure what happened on the server side but ugh. :/

This worked for me 🤷🏻‍♂️

**Comment by flikkr** [comment]
I'm facing this issue and tried several approaches unsuccessfully. Google provider is enabled in dashboard and I copied over my Android client id in the `Authorized Client IDs` field, and skip nonce check. I try doing login on my Android device using the code provided in [the README](https://pub.dev/packages/supabase_flutter#native-google-sign-in):

```
Future<AuthResponse> login() async {
  final GoogleSignIn googleSignIn = GoogleSignIn(
    clientId: Env.googleAuthAndroidClientId,
    serverClientId: Env.googleAuthWebClientId,
  );

  final googleUser = await googleSignIn.signIn();
  final googleAuth = await googleUser!.authentication;
  final accessToken = googleAuth.accessToken;
  final idToken = googleAuth.idToken;

  // TODO: cleanup
  if (accessToken == null) {
    throw 'No Access Token found.';
  }
  if (idToken == null) {
    throw 'No ID Token found.';
  }

  return Supabase.instance.client.auth.signInWithIdToken(
    provider: Provider.google,
    idToken: idToken,
    accessToken: accessToken,
  );
}
```

I feel like I'm running out of ideas as to what to do. I tried disabling/enabling provider to no success. Dashboard logs show this error.

> {"component":"api","error":"invalid request: Unacceptable audience in id_token","level":"info","method":"POST","msg":"invalid request: Unacceptable audience in id_token","path":"/token","referer":"http://localhost:3000","remote_addr":"218.212.109.252","time":"2023-12-06T16:38:21Z","timestamp":"2023-12-06T16:38:21Z"}

**Comment by BGM-109** [comment]
@dshukertjr same....here
